frontend/src/lib/Utils.tsx

```diff
@@ -292,8 +292,23 @@ export function generateGcsConsoleUri(gcsUri: string): string | undefined {
 
 const MINIO_URI_PREFIX = 'minio://';
 
-function generateArtifactUrl(source: string, bucket: string, key: string): string {
-  return `artifacts/get?source=${source}&bucket=${bucket}&key=${key}`;
+/**
+ * Generates the path component of the url to retrieve an artifact.
+ *
+ * @param source source of the artifact. Can be "minio", "s3", "http", "https", or "gcs".
+ * @param bucket bucket where the artifact is stored, value is assumed to be uri encoded.
+ * @param key path to the artifact, value is assumed to be uri encoded.
+ * @param peek number of characters or bytes to return. If not provided, the entire content of the artifact will be returned.
+ */
+export function generateArtifactUrl(
+  source: string,
+  bucket: string,
+  key: string,
+  peek?: number,
+): string {
+  return encodeURI(
+    `artifacts/get?source=${source}&bucket=${bucket}&key=${key}${peek ? `&peek=${peek}` : ''}`,
```

Comment:
i intentionally did not use buildQuery because of the encodeUriComponent. Because I think the keys and bucket etc from the S3 artifact is already encoded. 

In my prev PR, i created the consistent encode and decode because of this issue - double encoded in some cases. 

But let me make the change and try again.

Reply:
I see, if there's something blocking that, feel free to leave it for future improvements.

Reply:
ok. I did a test run, so far all the workflow artifacts and ml-metadata artifacts look fine too.